### PR TITLE
299 saving forms v2

### LINKED ISSUE #299: Part 2. Saving - Saving forms

### PR DESCRIPTION
**Pull request by kronken**
Closes #299

### CODE REVIEW

File: src/components/sidebar/routePathView/RoutePathView.tsx [hunk 1]
```diff
@@ -16,55 +28,125 @@ class RoutePathView extends React.Component<IRoutePathViewProps, IRoutePathViewS
...
+        } catch {
+            NotificationStore.addNotification({
+                message: 'Tallennus epäonnistui',
+                type: NotificationType.ERROR,
+            });
+        }
+        this.setState({ isLoading: false });
+    }
+
+    public onEdit = (property: string, value: any) => {
```
Comment (kronken) [inline src/components/sidebar/routePathView/RoutePathView.tsx]: Would onChange be a more natural word, it is used more often in this project

File: src/components/sidebar/routePathView/RoutePathView.tsx [hunk 2]
```diff
@@ -43,37 +47,68 @@ class RoutePathView extends React.Component<IRoutePathViewProps, IRoutePathViewS
...
+                type: NotificationType.ERROR,
+            });
+        }
+        this.setState({ isLoading: false });
+    }
+
+    public onChange = (property: string, value: any) => {
+        this.setState({
+            routePath: { ...this.state.routePath!, [property]: value },
+            hasModifications: true,
```
Comment (Divinare) [inline src/components/sidebar/routePathView/RoutePathView.tsx]: I think to see if we have notifications or not, we should have an initialState saved somewhere and current that we compare to find out if we have modifications

"What if user changes a field from xxx to xx, then back from xx to xxx"

File: src/util/ApiClient.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+enum RequestMethod {
+    POST= 'POST',
+    PUT= 'PUT',
+    DELETE= 'DELETE',
+}
+
+const API_URL = process.env.API_URL || 'http://localhost:3040';
+
+export default class ApiClient {
+    public async saveObject(entityName: entityNames, object: any) {
```
Comment (Divinare) [inline src/util/ApiClient.tsx]: editObject / updateObject

File: src/services/routePathService.ts [hunk 4]
```diff
@@ -29,6 +31,11 @@ export default class RoutePathService {
             return null;
         }
     }
+
+    public static async saveRoutePath(routePath: IRoutePath) {
+        const client = new ApiClient();
+        return await client.saveObject(entityNames.ROUTEPATH, routePath);
```
Comment (Divinare) [inline src/services/routePathService.ts]: apiClient

File: src/util/ApiClient.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import entityNames from '~/enums/entityNames';
+
+enum RequestMethod {
+    POST= 'POST',
+    PUT= 'PUT',
+    DELETE= 'DELETE',
+}
+
+const API_URL = process.env.API_URL || 'http://localhost:3040';
```
Comment (Divinare) [inline src/util/ApiClient.tsx]: add / to the end

File: src/components/sidebar/routePathView/RoutePathView.tsx (hunk 2 above)
Comment (kronken) [reply src/components/sidebar/routePathView/RoutePathView.tsx]: Maybe not required for a MVP, but added task so we remember it. https://github.com/HSLdevcom/jore-map-ui/issues/357

File: src/util/ApiClient.tsx (hunk 5 above)
Comment (kronken) [reply src/util/ApiClient.tsx]: Added it instead to row 36, thought that would be more natural

File: src/services/routePathService.ts (hunk 4 above)
Comment (kronken) [reply src/services/routePathService.ts]: fixed

File: src/components/sidebar/ViewHeader.tsx [hunk 6]
```diff
@@ -7,12 +7,15 @@ import * as s from './viewHeader.scss';
...
     header: string;
     children?: any;
+    onBeforeClosing?: Function;
 }
 
 class ViewHeader extends React.Component<IViewHeaderProps> {
     private closeSidebarView = () => {
-        const routesLink = routeBuilder.to(subSites.routes).toLink();
-        navigator.goTo(routesLink);
+        if (!this.props.onBeforeClosing || this.props.onBeforeClosing()) {
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/sidebar/ViewHeader.tsx]: fixed

**Pull request merged**